### ISSUE TITLE
Looping Other Vehicles in Custom Env

### ISSUE THREAD

**Issue by supperted825** [eleurent/highway-env#231]
Hello there,

I have created a custom environment as shown below:

![image](https://user-images.githubusercontent.com/74523278/133473624-549b0535-7988-492e-80ad-4cc54fed2cfa.png)

I haven't built my RL implementation yet, but I'm noticing some weird behaviour in the non-ego vehicles, particularly where the road network graph circles around. I have a > b > c > d > e > f > g > h > a... and so on. I want the non-ego vehicles to simply keep looping around the track.

The behaviours are ok from a to h, but the vehicles go crazy once they reach h, going off the track or driving in circles instead of just continuing with the loop again.

Is there somewhere that this behaviour can be changed, or perhaps is there a function specifically that I should modify to get the effect that I want?

**Comment by eleurent** [comment]
Hi @supperted825,
I don't really know what's going on, I think that you did everything right and it's supposed to work...
Would you mind sharing your code, so that I can investigate further?

**Comment by supperted825** [comment]
Hi there,

Thank you for the reply. I managed to work around the issue by splitting up the giant arc on the left of my track into two smaller arcs. My road network construction code is shown below (See: 7 - Circular Arc).

It's just a guess, but I think that when the arc was too big, the projected trajectory somehow points back to the start of the arc which is why they began looping only within that circle.

```
def _make_road(self) -> None:
        net = RoadNetwork()

        # Set Speed Limits for Road Sections - Straight, Turn20, Straight, Turn 15, Turn15, Straight, Turn25x2, Turn18
        speedlimits = [None, 10, 10, 10, 10, 10, 10, 10, 10]

        # Initialise First Lane
        lane = StraightLane([42, 0], [100, 0], line_types=(LineType.CONTINUOUS, LineType.STRIPED), width=5, speed_limit=speedlimits[1])
        self.lane = lane

        # Add Lanes to Road Network - Straight Section
        net.add_lane("a", "b", lane)
        net.add_lane("a", "b", StraightLane([42, 5], [100, 5], line_types=(LineType.STRIPED, LineType.CONTINUOUS), width=5, speed_limit=speedlimits[1]))

        # 2 - Circular Arc #1
        center1 = [100, -20]
        radii1 = 20
        net.add_lane("b", "c",
                        CircularLane(center1, radii1, np.deg2rad(90), np.deg2rad(-1), width=5,
                                    clockwise=False, line_types=(LineType.CONTINUOUS, LineType.NONE),
                                    speed_limit=speedlimits[2]))
        net.add_lane("b", "c",
                        CircularLane(center1, radii1+5, np.deg2rad(90), np.deg2rad(-1), width=5,
                                    clockwise=False, line_types=(LineType.STRIPED, LineType.CONTINUOUS),
                                    speed_limit=speedlimits[2]))

        # 3 - Vertical Straight
        net.add_lane("c", "d", StraightLane([120, -20], [120, -30],
                                line_types=(LineType.CONTINUOUS, LineType.NONE), width=5,
                                speed_limit=speedlimits[3]))
        net.add_lane("c", "d", StraightLane([125, -20], [125, -30],
                                line_types=(LineType.STRIPED, LineType.CONTINUOUS), width=5,
                                speed_limit=speedlimits[3]))

        # 4 - Circular Arc #2
        center2 = [105, -30]
        radii2 = 15
        net.add_lane("d", "e",
                        CircularLane(center2, radii2, np.deg2rad(0), np.deg2rad(-181), width=5,
                                    clockwise=False, line_types=(LineType.CONTINUOUS, LineType.NONE),
                                    speed_limit=speedlimits[4]))
        net.add_lane("d", "e",
                        CircularLane(center2, radii2+5, np.deg2rad(0), np.deg2rad(-181), width=5,
                                    clockwise=False, line_types=(LineType.STRIPED, LineType.CONTINUOUS),
                                    speed_limit=speedlimits[4]))

        # 5 - Circular Arc #3
        center3 = [70, -30]
        radii3 = 15
        net.add_lane("e", "f",
                        CircularLane(center3, radii3+5, np.deg2rad(0), np.deg2rad(136), width=5,
                                    clockwise=True, line_types=(LineType.CONTINUOUS, LineType.STRIPED),
                                    speed_limit=speedlimits[5]))
        net.add_lane("e", "f",
                        CircularLane(center3, radii3, np.deg2rad(0), np.deg2rad(137), width=5,
                                    clockwise=True, line_types=(LineType.NONE, LineType.CONTINUOUS),
                                    speed_limit=speedlimits[5]))

        # 6 - Slant
        net.add_lane("f", "g", StraightLane([55.7, -15.7], [35.7, -35.7],
                                            line_types=(LineType.CONTINUOUS, LineType.NONE), width=5,
                                            speed_limit=speedlimits[6]))
        net.add_lane("f", "g", StraightLane([59.3934, -19.2], [39.3934, -39.2],
                                            line_types=(LineType.STRIPED, LineType.CONTINUOUS), width=5,
                                            speed_limit=speedlimits[6]))

        # 7 - Circular Arc #4 - Bugs out when arc is too large, hence written in 2 sections
        center4 = [18.1, -18.1]
        radii4 = 25
        net.add_lane("g", "h",
                        CircularLane(center4, radii4, np.deg2rad(315), np.deg2rad(170), width=5,
                                    clockwise=False, line_types=(LineType.CONTINUOUS, LineType.NONE),
                                    speed_limit=speedlimits[7]))
        net.add_lane("g", "h",
                        CircularLane(center4, radii4+5, np.deg2rad(315), np.deg2rad(165), width=5,
                                    clockwise=False, line_types=(LineType.STRIPED, LineType.CONTINUOUS),
                                    speed_limit=speedlimits[7]))
        net.add_lane("h", "i",
                        CircularLane(center4, radii4, np.deg2rad(170), np.deg2rad(56), width=5,
                                    clockwise=False, line_types=(LineType.CONTINUOUS, LineType.NONE),
                                    speed_limit=speedlimits[7]))
        net.add_lane("h", "i",
                        CircularLane(center4, radii4+5, np.deg2rad(170), np.deg2rad(58), width=5,
                                    clockwise=False, line_types=(LineType.STRIPED, LineType.CONTINUOUS),
                                    speed_limit=speedlimits[7]))

        # 8 - Circular Arc #5 - Reconnects to Start
        center5 = [43.2, 23.4]
        radii5 = 18.5
        net.add_lane("i", "a",
                        CircularLane(center5, radii5+5, np.deg2rad(240), np.deg2rad(270), width=5,
                                    clockwise=True, line_types=(LineType.CONTINUOUS, LineType.STRIPED),
                                    speed_limit=speedlimits[8]))
        net.add_lane("i", "a",
                        CircularLane(center5, radii5, np.deg2rad(238), np.deg2rad(268), width=5,
                                    clockwise=True, line_types=(LineType.NONE, LineType.CONTINUOUS),
                                    speed_limit=speedlimits[8]))

        road = Road(network=net, np_random=self.np_random, record_history=self.config["show_trajectories"])
        self.road = road
```

**Comment by eleurent** [comment]
Thanks a lot! I looked into it, and here is what happens

* in the follow_road method, we try to find if a vehicle has finished its current road segment (lane) before switching to the next one
* this happens if the longitudinal coordinate of the vehicle (in the frame of the lane) is greater than the lane's length
* for CircularLanes, the longitidinal coordinate of any position is computed as follows: 
    longitudinal = distance to the center * angle to the start phase

Thus, whenever the angle to the start phase is greater than pi, it gets wrapped to -pi.  As an effet, the longitudinal coordinate of the vehicle jumps from 78 to -78, and the vehicle tries to drive around the full circle again (from l=-78) and never reaches the lane length coordinate l=90.

To avoid this, one possibility that you suggested is to never have an arc of more than180°.
However, it seems a bit unsatisfactory to me. I will try to think of a better solution, but I'm not sure there is one, at least without introducing a notion of time-continuity to the previous coordinate, to avoid this wrapping effect.

**Comment by eleurent** [comment]
Also: I really like your env, would you mind sharing it with everyone by opening a pull request once it is done?

**Comment by supperted825** [comment]
Hi @eleurent,

Thanks for shedding some light on the way it behaves. I think for now, the workaround is satisfactory for me.

No problem on sharing the env, I will share it once it's more complete.

**Comment by eleurent** [comment]
Hi @supperted825,

Unless you are already ready to share a complete version, would  you mind if I go ahead and push the version that you wrote above?
I think is already quite interesting, since it is enough to produce the result below

https://user-images.githubusercontent.com/1706935/133753435-8854eec2-c78c-4c3e-813a-9b6a70722e5f.mp4

And you can still add missing content with a PR later.

**Comment by supperted825** [comment]
Hi @eleurent,

Yes please do. Apologies for the lack of an update, some other factors have put my work on the env on hold for now.

**Comment by eleurent** [comment]
No worries at all, I understand!
I'll make sure you are properly credited for this work, and I would love to see your updates if you have some time to work on it again in the future.

**Comment by eleurent** [comment]
I merged my branch, which includes:
- a simple version of the environment (a single other vehicle)
- a training script with SB3's PPO
- a new (optional) discrete action space (i.e. discretized continuous actions, and not discrete meta actions) 
- updated documentation

This should be sufficient for a new release :)

Again, many thanks for this valuable addition, and feel free to open a PR whenever you want to make some changes!
